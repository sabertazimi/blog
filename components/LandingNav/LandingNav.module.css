.nav {
  @apply fixed left-0 top-0 z-20 flex h-full w-full flex-col items-center justify-center;
}

.banner {
  @apply fixed left-0 top-0 z-10 block h-full w-full bg-black/50;
  @apply backdrop-brightness-110 backdrop-contrast-125 backdrop-saturate-150;
}

.icon {
  @apply fixed left-12 top-12 z-100 cursor-pointer border-none bg-transparent text-lg font-extrabold text-light md:text-4xl;
}
