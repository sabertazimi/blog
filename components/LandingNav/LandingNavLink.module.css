.link {
  @apply relative mx-6 my-0 block cursor-pointer px-0 pt-0 pb-3;
}

.link a {
  @apply text-4xl font-extrabold leading-none text-light md:text-7xl;
}

.link::after {
  @apply absolute left-0 top-full h-2 w-full;
  @apply bg-gradient-primary rounded-lg;

  content: '';
}

.tooltip {
  @apply mb-16;
}
