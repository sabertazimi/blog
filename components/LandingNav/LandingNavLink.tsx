import type { ReactNode } from 'react'
import Link from '@/components/Link'
import { Bounce } from '@/components/Motion'
import Tooltip from '@/components/Tooltip'
import styles from './LandingNavLink.module.css'

interface Props {
  title: string
  href: string
  children: ReactNode
}

function LandingNavLink({ title, href, children }: Props): JSX.Element {
  return (
    <Tooltip className={styles.tooltip} placement="left" title={title}>
      <Bounce>
        <span className={styles.link} role="link">
          <Link href={href}>{children}</Link>
        </span>
      </Bounce>
    </Tooltip>
  )
}

export default LandingNavLink
