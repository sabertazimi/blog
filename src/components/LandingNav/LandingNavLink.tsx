import { Tooltip } from 'antd';
import classNames from 'classnames';
import { Link } from 'gatsby';
import type { ReactNode } from 'react';
import React from 'react';

interface Props {
  title: string;
  to: string;
  children: ReactNode;
}

const LandingNavLink = ({ title, to, children }: Props): JSX.Element => (
  <Tooltip className="mb-16" placement="left" title={title}>
    <Link to={to} role="link">
      <span
        className={classNames(
          'block px-0 pt-0 pb-3 mx-6 my-0 border-solid border-b-6 border-light',
          'text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl',
          'transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110'
        )}
      >
        {children}
      </span>
    </Link>
  </Tooltip>
);

export default LandingNavLink;
