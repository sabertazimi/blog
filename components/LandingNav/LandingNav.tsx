'use client'

import type { Route } from '@/types'
import type { MotionProps } from 'framer-motion'
import type { KeyboardEvent } from 'react'
import { Hamburger } from '@/components/Icons'
import { Switch } from '@/components/Motion'
import { routes as defaultRoutes } from '@/config'
import { useCallback, useState } from 'react'
import styles from './LandingNav.module.css'
import LandingNavLink from './LandingNavLink'

interface Props {
  routes?: Route[]
}

const navVariants: MotionProps['variants'] = {
  open: {
    x: 0,
    opacity: 1,
  },
  close: {
    x: '-100%',
    opacity: 0,
  },
}

const navTransition: MotionProps['transition'] = {
  type: 'tween',
  ease: 'easeInOut',
  duration: 0.4,
}

const bannerVariants: MotionProps['variants'] = {
  open: {
    x: 0,
  },
  close: {
    x: '-100%',
  },
}

const bannerTransition: MotionProps['transition'] = {
  type: 'tween',
  ease: 'circInOut',
  duration: 0.4,
}

function LandingNav({ routes = defaultRoutes }: Props): JSX.Element {
  const [expanded, setExpanded] = useState(false)

  const handleClick = useCallback(() => {
    setExpanded(expanded => !expanded)
  }, [])

  const handleKeyDown = useCallback((event: KeyboardEvent<HTMLSpanElement>) => {
    if (event.key === 'Enter')
      setExpanded(expanded => !expanded)
  }, [])

  return (
    <>
      <Hamburger
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        className={styles.icon}
        data-testid="hamburger-button"
      />
      <Switch
        className={styles.nav}
        role="navigation"
        open={expanded}
        variants={navVariants}
        transition={navTransition}
      >
        {routes.map(route => (
          <LandingNavLink key={route.id} title={route.title} href={route.path}>
            {route.name}
          </LandingNavLink>
        ))}
      </Switch>
      <Switch
        className={styles.banner}
        role="banner"
        open={expanded}
        variants={bannerVariants}
        transition={bannerTransition}
      />
    </>
  )
}

export default LandingNav
