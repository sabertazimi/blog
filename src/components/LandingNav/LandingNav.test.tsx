import { Routes } from '@config';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { act, create } from 'react-test-renderer';
import LandingNav from './LandingNav';

describe('LandingNav', () => {
  test('should render routes correctly (snapshot)', () => {
    const renderer = create(<LandingNav routes={Routes} />);
    const tree = renderer.toJSON();
    expect(tree).toMatchSnapshot();

    const instance = renderer.root;
    const navButton = instance.find(
      node => node.type === 'div' && node.props.role === 'button'
    );
    act(() => navButton.props.onClick());

    const expandTree = renderer.toJSON();
    expect(expandTree).toMatchSnapshot();
  });

  test('should render route with correct structure', () => {
    render(<LandingNav routes={Routes} />);
    const nav = screen.getByRole('navigation');
    const navLinks = screen.getAllByRole('link');
    const navButton = screen.getByRole('button');
    const icon = screen.getByRole('img');

    expect(nav).toBeInTheDocument();
    navLinks.forEach(navLink => expect(navLink).toBeInTheDocument());
    expect(navButton).toBeInTheDocument();
    expect(icon).toBeInTheDocument();

    navLinks.forEach(navLink => expect(nav).toContainElement(navLink));
    expect(navButton).toContainElement(icon);
  });

  test('should expanded when clicked', () => {
    render(<LandingNav routes={Routes} />);
    const nav = screen.getByRole('navigation');
    const overlay = screen.getByRole('banner');
    const navButton = screen.getByRole('button');

    fireEvent.click(navButton);
    expect(nav).toHaveClass('translate-x-0');
    expect(overlay).toHaveClass('bg-opacity-80');
  });
});
