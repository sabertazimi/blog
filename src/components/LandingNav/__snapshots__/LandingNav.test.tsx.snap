// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingNav should render routes correctly (snapshot) 1`] = `
Array [
  <nav
    className="fixed top-0 left-0 flex flex-col items-center justify-center w-full h-full transition-all transform-gpu duration-500 z-100 -translate-x-full"
    role="navigation"
  >
    <a
      href="/posts"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        Posts
      </span>
    </a>
    <a
      href="/tags"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        Tags
      </span>
    </a>
    <a
      href="/books"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        Books
      </span>
    </a>
    <a
      href="/about"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        About
      </span>
    </a>
  </nav>,
  <div
    className="fixed bg-transparent border-none cursor-pointer top-12 left-12 z-100"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      className="anticon text-lg font-extrabold md:text-4xl text-light"
      role="img"
    >
      <svg
        fill="currentColor"
        height="2em"
        p-id="7522"
        viewBox="0 0 1024 1024"
        width="2em"
      >
        <path
          d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zM904 784H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zM904 472H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
          p-id="7523"
        />
      </svg>
    </span>
  </div>,
  <div
    className="fixed top-0 left-0 z-10 block w-full h-full transition-all transform-gpu bg-black bg-opacity-0"
    role="banner"
  />,
]
`;

exports[`LandingNav should render routes correctly (snapshot) 2`] = `
Array [
  <nav
    className="fixed top-0 left-0 flex flex-col items-center justify-center w-full h-full transition-all transform-gpu duration-500 z-100 translate-x-0"
    role="navigation"
  >
    <a
      href="/posts"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        Posts
      </span>
    </a>
    <a
      href="/tags"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        Tags
      </span>
    </a>
    <a
      href="/books"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        Books
      </span>
    </a>
    <a
      href="/about"
      role="link"
    >
      <span
        className="block px-0 pt-0 pb-3 mx-6 mb-16 border-solid border-b-6 border-light text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110"
      >
        About
      </span>
    </a>
  </nav>,
  <div
    className="fixed bg-transparent border-none cursor-pointer top-12 left-12 z-100"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      className="anticon text-lg font-extrabold md:text-4xl text-light"
      role="img"
    >
      <svg
        fill="currentColor"
        height="2em"
        p-id="8392"
        viewBox="0 0 1024 1024"
        width="2em"
      >
        <path
          d="M563.8 512l262.5-312.9c4.4-5.2 0.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9c-4.4 5.2-0.7 13.1 6.1 13.1h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
          p-id="8393"
        />
      </svg>
    </span>
  </div>,
  <div
    className="fixed top-0 left-0 z-10 block w-full h-full transition-all transform-gpu bg-black bg-opacity-80"
    role="banner"
  />,
]
`;
